atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java

```diff
@@ -16,79 +16,54 @@
+import java.util.HashSet;
+import java.util.Set;
 
-/** Represents a persisted server configuration. */
-public class Configuration {
-
-  private final long index;
-  private final long term;
-  private final long time;
-  private final Collection<RaftMember> members;
+/**
+ * Represents a persisted server configuration.
+ *
+ * @param index The index is the index of the {@link
+ *     io.atomix.raft.storage.log.entry.ConfigurationEntry ConfigurationEntry} which resulted in
+ *     this configuration.
+ * @param term The term is the term of the leader at the time the configuration change was
+ *     committed.
+ * @param time The time at which the configuration was committed.
+ * @param newMembers The cluster membership for this configuration.
+ * @param oldMembers The cluster membership for the previous configuration.
+ */
+public record Configuration(
+    long index,
+    long term,
+    long time,
+    Collection<RaftMember> newMembers,
+    Collection<RaftMember> oldMembers) {
 
-  public Configuration(
-      final long index, final long term, final long time, final Collection<RaftMember> members) {
+  public Configuration {
     checkArgument(time > 0, "time must be positive");
-    checkNotNull(members, "members cannot be null");
-    this.index = index;
-    this.term = term;
-    this.time = time;
-    this.members = members;
+    checkNotNull(newMembers, "members cannot be null");
+    checkNotNull(oldMembers, "oldMembers cannot be null");
   }
 
-  /**
-   * Returns the configuration index.
-   *
-   * <p>The index is the index of the {@link io.atomix.raft.storage.log.entry.ConfigurationEntry
-   * ConfigurationEntry} which resulted in this configuration.
-   *
-   * @return The configuration index.
-   */
-  public long index() {
-    return index;
-  }
-
-  /**
-   * Returns the configuration term.
-   *
-   * <p>The term is the term of the leader at the time the configuration change was committed.
-   *
-   * @return The configuration term.
-   */
-  public long term() {
-    return term;
-  }
-
-  /**
-   * Returns the configuration time.
-   *
-   * @return The time at which the configuration was committed.
-   */
-  public long time() {
-    return time;
+  public Configuration(
+      final long index, final long term, final long time, final Collection<RaftMember> members) {
+    this(index, term, time, members, Collections.emptyList());
   }
 
-  /**
-   * Returns the cluster membership for this configuration.
-   *
-   * @return The cluster membership.
-   */
-  public Collection<RaftMember> members() {
-    return members;
+  public boolean requiresJointConsensus() {
+    return !oldMembers.isEmpty();
   }
 
-  @Override
-  public String toString() {
-    return toStringHelper(this)
-        .add("index", index)
-        .add("time", new TimestampPrinter(time))
-        .add("members", members)
-        .toString();
+  public Set<RaftMember> allMembers() {
+    return new HashSet<>(newMembers) {
```

Comment:
:wrench: Double brace initialization is considered an anti-pattern.

Reply:
:disappointed: I just discovered it for the first time. It does look odd though, I'll remove it :+1: